File: mobile/lib/data/models/features.dart
import 'user.dart';
import 'category.dart';

/// =============================================================================
/// AUTO-BID MODELS
/// =============================================================================

/// AutoBid represents an auto-bid configuration
class AutoBid {
  final String id;
  final String auctionId;
  final String userId;
  final double maxAmount;
  final double? currentBidAmount;
  final bool isActive;
  final DateTime createdAt;
  final DateTime updatedAt;
  final DateTime? deactivatedAt;
  final String? deactivationReason;
  
  // Extended
  final String? auctionTitle;
  final double? currentPrice;
  final double? startingPrice;
  final DateTime? endTime;

  AutoBid({
    required this.id,
    required this.auctionId,
    required this.userId,
    required this.maxAmount,
    this.currentBidAmount,
    required this.isActive,
    required this.createdAt,
    required this.updatedAt,
    this.deactivatedAt,
    this.deactivationReason,
    this.auctionTitle,
    this.currentPrice,
    this.startingPrice,
    this.endTime,
  });

  factory AutoBid.fromJson(Map<String, dynamic> json) {
    return AutoBid(
      id: json['id'] as String,
      auctionId: json['auction_id'] as String,
      userId: json['user_id'] as String,
      maxAmount: (json['max_amount'] as num).toDouble(),
      currentBidAmount: json['current_bid_amount'] != null
          ? (json['current_bid_amount'] as num).toDouble()
          : null,
      isActive: json['is_active'] as bool? ?? false,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      deactivatedAt: json['deactivated_at'] != null
          ? DateTime.parse(json['deactivated_at'] as String)
          : null,
      deactivationReason: json['deactivation_reason'] as String?,
      auctionTitle: json['auction_title'] as String?,
      currentPrice: json['current_price'] != null
          ? (json['current_price'] as num).toDouble()
          : null,
      startingPrice: json['starting_price'] != null
          ? (json['starting_price'] as num).toDouble()
          : null,
      endTime: json['end_time'] != null
          ? DateTime.parse(json['end_time'] as String)
          : null,
    );
  }
}

/// AutoBidResponse after setting auto-bid
class AutoBidResponse {
  final AutoBid? autoBid;
  final String message;
  final double currentBid;
  final double nextBid;
  final int bidsPlaced;
  final bool isHighBidder;

  AutoBidResponse({
    this.autoBid,
    required this.message,
    required this.currentBid,
    required this.nextBid,
    required this.bidsPlaced,
    required this.isHighBidder,
  });

  factory AutoBidResponse.fromJson(Map<String, dynamic> json) {
    return AutoBidResponse(
      autoBid: json['auto_bid'] != null
          ? AutoBid.fromJson(json['auto_bid'] as Map<String, dynamic>)
          : null,
      message: json['message'] as String? ?? '',
      currentBid: (json['current_bid'] as num?)?.toDouble() ?? 0,
      nextBid: (json['next_bid'] as num?)?.toDouble() ?? 0,
      bidsPlaced: json['bids_placed'] as int? ?? 0,
      isHighBidder: json['is_high_bidder'] as bool? ?? false,
    );
  }
}

/// =============================================================================
/// SAVED SEARCH MODELS
/// =============================================================================

/// SavedSearch represents a user's saved search
class SavedSearch {
  final String id;
  final String userId;
  final String name;
  final String? searchQuery;
  final String? categoryId;
  final String? townId;
  final double? minPrice;
  final double? maxPrice;
  final List<String>? keywords;
  final String? condition;
  final bool notifyNewListings;
  final bool notifyPriceDrops;
  final bool notifyEmail;
  final bool notifyPush;
  final int matchCount;
  final bool isActive;
  final DateTime? lastNotifiedAt;
  final DateTime createdAt;
  final DateTime updatedAt;
  
  // Joined
  final Category? category;
  final Town? town;

  SavedSearch({
    required this.id,
    required this.userId,
    required this.name,
    this.searchQuery,
    this.categoryId,
    this.townId,
    this.minPrice,
    this.maxPrice,
    this.keywords,
    this.condition,
    this.notifyNewListings = true,
    this.notifyPriceDrops = false,
    this.notifyEmail = true,
    this.notifyPush = true,
    this.matchCount = 0,
    this.isActive = true,
    this.lastNotifiedAt,
    required this.createdAt,
    required this.updatedAt,
    this.category,
    this.town,
  });

  factory SavedSearch.fromJson(Map<String, dynamic> json) {
    return SavedSearch(
      id: json['id'] as String,
      userId: json['user_id'] as String? ?? '',
      name: json['name'] as String? ?? 'Unnamed Search',
      searchQuery: json['search_query'] as String?,
      categoryId: json['category_id'] as String?,
      townId: json['town_id'] as String?,
      minPrice: json['min_price'] != null ? (json['min_price'] as num).toDouble() : null,
      maxPrice: json['max_price'] != null ? (json['max_price'] as num).toDouble() : null,
      keywords: json['keywords'] != null ? List<String>.from(json['keywords'] as List) : null,
      condition: json['condition'] as String?,
      notifyNewListings: json['notify_new_listings'] as bool? ?? true,
      notifyPriceDrops: json['notify_price_drops'] as bool? ?? false,
      notifyEmail: json['notify_email'] as bool? ?? true,
      notifyPush: json['notify_push'] as bool? ?? true,
      matchCount: json['match_count'] as int? ?? 0,
      isActive: json['is_active'] as bool? ?? true,
      lastNotifiedAt: json['last_notified_at'] != null
          ? DateTime.parse(json['last_notified_at'] as String)
          : null,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : DateTime.now(),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'] as String)
          : DateTime.now(),
      category: json['category'] != null
          ? Category.fromJson(json['category'] as Map<String, dynamic>)
          : null,
      town: json['town'] != null
          ? Town.fromJson(json['town'] as Map<String, dynamic>)
          : null,
    );
  }
}

/// =============================================================================
/// PROMOTION MODELS
/// =============================================================================

/// PromotionPricing for available promotion options
class PromotionPricing {
  final String id;
  final String name;
  final String promotionType;
  final int durationHours;
  final double price;
  final double boostMultiplier;
  final String description;
  final bool isActive;
  final String? townId;

  PromotionPricing({
    required this.id,
    required this.name,
    required this.promotionType,
    required this.durationHours,
    required this.price,
    required this.boostMultiplier,
    required this.description,
    this.isActive = true,
    this.townId,
  });

  factory PromotionPricing.fromJson(Map<String, dynamic> json) {
    return PromotionPricing(
      id: json['id'] as String,
      name: json['name'] as String? ?? '',
      promotionType: json['promotion_type'] as String? ?? 'featured',
      durationHours: json['duration_hours'] as int? ?? 24,
      price: (json['price'] as num?)?.toDouble() ?? 0,
      boostMultiplier: (json['boost_multiplier'] as num?)?.toDouble() ?? 1.0,
      description: json['description'] as String? ?? '',
      isActive: json['is_active'] as bool? ?? true,
      townId: json['town_id'] as String?,
    );
  }

  String get durationFormatted {
    if (durationHours < 24) return '$durationHours hours';
    if (durationHours == 24) return '1 day';
    if (durationHours < 168) return '${durationHours ~/ 24} days';
    if (durationHours == 168) return '1 week';
    return '${durationHours ~/ 168} weeks';
  }
}

/// PromotedAuction representing active promotion
class PromotedAuction {
  final String id;
  final String auctionId;
  final String userId;
  final String promotionType;
  final String? townId;
  final DateTime startsAt;
  final DateTime endsAt;
  final double amountPaid;
  final bool isActive;
  final int impressions;
  final int clicks;
  final double boostMultiplier;
  final String paymentStatus;
  final DateTime createdAt;

  PromotedAuction({
    required this.id,
    required this.auctionId,
    required this.userId,
    required this.promotionType,
    this.townId,
    required this.startsAt,
    required this.endsAt,
    required this.amountPaid,
    this.isActive = true,
    this.impressions = 0,
    this.clicks = 0,
    this.boostMultiplier = 1.0,
    this.paymentStatus = 'pending',
    required this.createdAt,
  });

  factory PromotedAuction.fromJson(Map<String, dynamic> json) {
    return PromotedAuction(
      id: json['id'] as String,
      auctionId: json['auction_id'] as String,
      userId: json['user_id'] as String,
      promotionType: json['promotion_type'] as String? ?? 'featured',
      townId: json['town_id'] as String?,
      startsAt: DateTime.parse(json['starts_at'] as String),
      endsAt: DateTime.parse(json['ends_at'] as String),
      amountPaid: (json['amount_paid'] as num?)?.toDouble() ?? 0,
      isActive: json['is_active'] as bool? ?? true,
      impressions: json['impressions'] as int? ?? 0,
      clicks: json['clicks'] as int? ?? 0,
      boostMultiplier: (json['boost_multiplier'] as num?)?.toDouble() ?? 1.0,
      paymentStatus: json['payment_status'] as String? ?? 'pending',
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  bool get isExpired => DateTime.now().isAfter(endsAt);
  Duration get timeRemaining => endsAt.difference(DateTime.now());
}

/// =============================================================================
/// USER REPUTATION MODELS
/// =============================================================================

/// UserReputation with detailed trust info
class UserReputation {
  final String userId;
  final String username;
  final String fullName;
  final String? avatarUrl;
  final double rating;
  final int ratingCount;
  final int completedAuctions;
  final String badgeLevel; // none, bronze, silver, gold
  final bool isTrustedSeller;
  final bool isVerified;
  final bool isFastResponder;
  final double completionRate;
  final int totalTransactions;
  final int successfulTransactions;
  final DateTime memberSince;
  final List<String> badges;

  UserReputation({
    required this.userId,
    required this.username,
    required this.fullName,
    this.avatarUrl,
    this.rating = 0,
    this.ratingCount = 0,
    this.completedAuctions = 0,
    this.badgeLevel = 'none',
    this.isTrustedSeller = false,
    this.isVerified = false,
    this.isFastResponder = false,
    this.completionRate = 100,
    this.totalTransactions = 0,
    this.successfulTransactions = 0,
    required this.memberSince,
    this.badges = const [],
  });

  factory UserReputation.fromJson(Map<String, dynamic> json) {
    return UserReputation(
      userId: json['user_id'] as String,
      username: json['username'] as String? ?? '',
      fullName: json['full_name'] as String? ?? '',
      avatarUrl: json['avatar_url'] as String?,
      rating: (json['rating'] as num?)?.toDouble() ?? 0,
      ratingCount: json['rating_count'] as int? ?? 0,
      completedAuctions: json['completed_auctions'] as int? ?? 0,
      badgeLevel: json['badge_level'] as String? ?? 'none',
      isTrustedSeller: json['is_trusted_seller'] as bool? ?? false,
      isVerified: json['is_verified'] as bool? ?? false,
      isFastResponder: json['is_fast_responder'] as bool? ?? false,
      completionRate: (json['completion_rate'] as num?)?.toDouble() ?? 100,
      totalTransactions: json['total_transactions'] as int? ?? 0,
      successfulTransactions: json['successful_transactions'] as int? ?? 0,
      memberSince: json['member_since'] != null
          ? DateTime.parse(json['member_since'] as String)
          : DateTime.now(),
      badges: json['badges'] != null
          ? List<String>.from(json['badges'] as List)
          : [],
    );
  }

  String get ratingDisplay => rating.toStringAsFixed(1);
  bool get isTopSeller => rating >= 4.5 && completedAuctions >= 10;
  
  String get badgeEmoji {
    switch (badgeLevel) {
      case 'gold': return '🥇';
      case 'silver': return '🥈';
      case 'bronze': return '🥉';
      default: return '';
    }
  }
}

/// UserRating represents a review
class UserRating {
  final String id;
  final String raterId;
  final String ratedUserId;
  final String? auctionId;
  final int rating;
  final int? communicationRating;
  final int? accuracyRating;
  final int? speedRating;
  final String? review;
  final String role; // buyer, seller
  final bool wouldRecommend;
  final DateTime createdAt;
  final User? rater;

  UserRating({
    required this.id,
    required this.raterId,
    required this.ratedUserId,
    this.auctionId,
    required this.rating,
    this.communicationRating,
    this.accuracyRating,
    this.speedRating,
    this.review,
    required this.role,
    this.wouldRecommend = true,
    required this.createdAt,
    this.rater,
  });

  factory UserRating.fromJson(Map<String, dynamic> json) {
    return UserRating(
      id: json['id'] as String,
      raterId: json['rater_id'] as String? ?? '',
      ratedUserId: json['rated_user_id'] as String? ?? '',
      auctionId: json['auction_id'] as String?,
      rating: json['rating'] as int? ?? 5,
      communicationRating: json['communication_rating'] as int?,
      accuracyRating: json['accuracy_rating'] as int?,
      speedRating: json['speed_rating'] as int?,
      review: json['review'] as String?,
      role: json['role'] as String? ?? 'buyer',
      wouldRecommend: json['would_recommend'] as bool? ?? true,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : DateTime.now(),
      rater: json['rater'] != null
          ? User.fromJson(json['rater'] as Map<String, dynamic>)
          : null,
    );
  }
}

/// =============================================================================
/// TOWN LEADERBOARD MODELS
/// =============================================================================

/// LeaderboardEntry for town rankings
class LeaderboardEntry {
  final int rank;
  final String userId;
  final String username;
  final String fullName;
  final String? avatarUrl;
  final String badgeLevel;
  final double score;
  final int metricValue;

  LeaderboardEntry({
    required this.rank,
    required this.userId,
    required this.username,
    required this.fullName,
    this.avatarUrl,
    this.badgeLevel = 'none',
    required this.score,
    required this.metricValue,
  });

  factory LeaderboardEntry.fromJson(Map<String, dynamic> json) {
    return LeaderboardEntry(
      rank: json['rank'] as int? ?? 0,
      userId: json['user_id'] as String,
      username: json['username'] as String? ?? '',
      fullName: json['full_name'] as String? ?? '',
      avatarUrl: json['avatar_url'] as String?,
      badgeLevel: json['badge_level'] as String? ?? 'none',
      score: (json['score'] as num?)?.toDouble() ?? 0,
      metricValue: json['metric_value'] as int? ?? 0,
    );
  }

  String get badgeEmoji {
    switch (badgeLevel) {
      case 'gold': return '🥇';
      case 'silver': return '🥈';
      case 'bronze': return '🥉';
      default: return '';
    }
  }
}

/// TownLeaderboard with all entries
class TownLeaderboard {
  final String townId;
  final String townName;
  final String leaderboardType;
  final String period;
  final List<LeaderboardEntry> entries;
  final DateTime calculatedAt;

  TownLeaderboard({
    required this.townId,
    required this.townName,
    required this.leaderboardType,
    required this.period,
    required this.entries,
    required this.calculatedAt,
  });

  factory TownLeaderboard.fromJson(Map<String, dynamic> json) {
    return TownLeaderboard(
      townId: json['town_id'] as String,
      townName: json['town_name'] as String? ?? '',
      leaderboardType: json['leaderboard_type'] as String? ?? 'top_sellers',
      period: json['period'] as String? ?? 'monthly',
      entries: (json['entries'] as List?)
              ?.map((e) => LeaderboardEntry.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      calculatedAt: json['calculated_at'] != null
          ? DateTime.parse(json['calculated_at'] as String)
          : DateTime.now(),
    );
  }
}

/// TownStats aggregated statistics
class TownStats {
  final String townId;
  final String townName;
  final int activeAuctions;
  final int totalAuctions;
  final int totalUsers;
  final int activeSellers;
  final double totalSalesValue;
  final double avgAuctionPrice;
  final String? topCategoryId;
  final Category? topCategory;
  final DateTime calculatedAt;

  TownStats({
    required this.townId,
    required this.townName,
    this.activeAuctions = 0,
    this.totalAuctions = 0,
    this.totalUsers = 0,
    this.activeSellers = 0,
    this.totalSalesValue = 0,
    this.avgAuctionPrice = 0,
    this.topCategoryId,
    this.topCategory,
    required this.calculatedAt,
  });

  factory TownStats.fromJson(Map<String, dynamic> json) {
    return TownStats(
      townId: json['town_id'] as String,
      townName: json['town_name'] as String? ?? '',
      activeAuctions: json['active_auctions'] as int? ?? 0,
      totalAuctions: json['total_auctions'] as int? ?? 0,
      totalUsers: json['total_users'] as int? ?? 0,
      activeSellers: json['active_sellers'] as int? ?? 0,
      totalSalesValue: (json['total_sales_value'] as num?)?.toDouble() ?? 0,
      avgAuctionPrice: (json['avg_auction_price'] as num?)?.toDouble() ?? 0,
      topCategoryId: json['top_category_id'] as String?,
      topCategory: json['top_category'] != null
          ? Category.fromJson(json['top_category'] as Map<String, dynamic>)
          : null,
      calculatedAt: json['calculated_at'] != null
          ? DateTime.parse(json['calculated_at'] as String)
          : DateTime.now(),
    );
  }
}

File: backend/internal/models/features.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// =============================================================================
// AUTO-BID MODELS
// =============================================================================

// AutoBid represents an auto-bid configuration
type AutoBid struct {
	ID                 uuid.UUID  `json:"id"`
	AuctionID          uuid.UUID  `json:"auction_id"`
	UserID             uuid.UUID  `json:"user_id"`
	MaxAmount          float64    `json:"max_amount"`
	CurrentBidAmount   *float64   `json:"current_bid_amount,omitempty"`
	IsActive           bool       `json:"is_active"`
	CreatedAt          time.Time  `json:"created_at"`
	UpdatedAt          time.Time  `json:"updated_at"`
	DeactivatedAt      *time.Time `json:"deactivated_at,omitempty"`
	DeactivationReason *string    `json:"deactivation_reason,omitempty"`
}

// CreateAutoBidRequest for setting up auto-bid
type CreateAutoBidRequest struct {
	MaxAmount float64 `json:"max_amount" binding:"required,min=0.01"`
}

// AutoBidResponse after creating/updating auto-bid
type AutoBidResponse struct {
	AutoBid      *AutoBid `json:"auto_bid"`
	Message      string   `json:"message"`
	CurrentBid   float64  `json:"current_bid"`
	NextBid      float64  `json:"next_bid"`
	BidsPlaced   int      `json:"bids_placed"`
	IsHighBidder bool     `json:"is_high_bidder"`
}

// =============================================================================
// SAVED SEARCH & ALERT MODELS
// =============================================================================

// SavedSearch represents a user's saved search
type SavedSearch struct {
	ID                uuid.UUID  `json:"id"`
	UserID            uuid.UUID  `json:"user_id"`
	Name              string     `json:"name"`
	SearchQuery       *string    `json:"search_query,omitempty"`
	CategoryID        *uuid.UUID `json:"category_id,omitempty"`
	TownID            *uuid.UUID `json:"town_id,omitempty"`
	MinPrice          *float64   `json:"min_price,omitempty"`
	MaxPrice          *float64   `json:"max_price,omitempty"`
	Keywords          []string   `json:"keywords,omitempty"`
	Condition         *string    `json:"condition,omitempty"`
	NotifyNewListings bool       `json:"notify_new_listings"`
	NotifyPriceDrops  bool       `json:"notify_price_drops"`
	NotifyEmail       bool       `json:"notify_email"`
	NotifyPush        bool       `json:"notify_push"`
	MatchCount        int        `json:"match_count"`
	IsActive          bool       `json:"is_active"`
	LastNotifiedAt    *time.Time `json:"last_notified_at,omitempty"`
	CreatedAt         time.Time  `json:"created_at"`
	UpdatedAt         time.Time  `json:"updated_at"`

	// Joined
	Category *Category `json:"category,omitempty"`
	Town     *Town     `json:"town,omitempty"`
}

// CreateSavedSearchRequest for saving a search
type CreateSavedSearchRequest struct {
	Name              string     `json:"name" binding:"required,min=1,max=100"`
	SearchQuery       *string    `json:"search_query"`
	CategoryID        *uuid.UUID `json:"category_id"`
	TownID            *uuid.UUID `json:"town_id"`
	MinPrice          *float64   `json:"min_price"`
	MaxPrice          *float64   `json:"max_price"`
	Keywords          []string   `json:"keywords"`
	Condition         *string    `json:"condition"`
	NotifyNewListings bool       `json:"notify_new_listings"`
	NotifyPriceDrops  bool       `json:"notify_price_drops"`
	NotifyEmail       bool       `json:"notify_email"`
	NotifyPush        bool       `json:"notify_push"`
}

// SearchAlert represents a notification for a saved search match
type SearchAlert struct {
	ID            uuid.UUID `json:"id"`
	SavedSearchID uuid.UUID `json:"saved_search_id"`
	AuctionID     uuid.UUID `json:"auction_id"`
	UserID        uuid.UUID `json:"user_id"`
	AlertType     string    `json:"alert_type"` // new_match, price_drop
	WasRead       bool      `json:"was_read"`
	CreatedAt     time.Time `json:"created_at"`

	// Joined
	Auction *Auction `json:"auction,omitempty"`
}

// =============================================================================
// PROMOTION MODELS
// =============================================================================

// PromotedAuction represents a boosted auction
type PromotedAuction struct {
	ID              uuid.UUID  `json:"id"`
	AuctionID       uuid.UUID  `json:"auction_id"`
	UserID          uuid.UUID  `json:"user_id"`
	PromotionType   string     `json:"promotion_type"` // featured, boosted, pinned, highlighted
	TownID          *uuid.UUID `json:"town_id,omitempty"`
	StartsAt        time.Time  `json:"starts_at"`
	EndsAt          time.Time  `json:"ends_at"`
	AmountPaid      float64    `json:"amount_paid"`
	IsActive        bool       `json:"is_active"`
	Impressions     int        `json:"impressions"`
	Clicks          int        `json:"clicks"`
	BoostMultiplier float64    `json:"boost_multiplier"`
	PaymentStatus   string     `json:"payment_status"`
	CreatedAt       time.Time  `json:"created_at"`
}

// PromotionPricing represents available promotion options
type PromotionPricing struct {
	ID              uuid.UUID  `json:"id"`
	Name            string     `json:"name"`
	PromotionType   string     `json:"promotion_type"`
	DurationHours   int        `json:"duration_hours"`
	Price           float64    `json:"price"`
	BoostMultiplier float64    `json:"boost_multiplier"`
	Description     string     `json:"description"`
	IsActive        bool       `json:"is_active"`
	TownID          *uuid.UUID `json:"town_id,omitempty"`
}

// CreatePromotionRequest for promoting an auction
type CreatePromotionRequest struct {
	PricingID uuid.UUID `json:"pricing_id" binding:"required"`
}

// PromotionResponse after creating promotion
type PromotionResponse struct {
	Promotion *PromotedAuction `json:"promotion"`
	Message   string           `json:"message"`
	ExpiresAt time.Time        `json:"expires_at"`
}

// =============================================================================
// SLOT PURCHASE MODELS
// =============================================================================

// SlotPurchase represents a queue skip purchase
type SlotPurchase struct {
	ID                   uuid.UUID  `json:"id"`
	UserID               uuid.UUID  `json:"user_id"`
	CategoryID           uuid.UUID  `json:"category_id"`
	TownID               uuid.UUID  `json:"town_id"`
	AuctionID            *uuid.UUID `json:"auction_id,omitempty"`
	AmountPaid           float64    `json:"amount_paid"`
	OriginalWaitPosition int        `json:"original_wait_position"`
	PurchaseType         string     `json:"purchase_type"` // skip_queue, extra_slot
	Status               string     `json:"status"`        // pending, completed, refunded
	CreatedAt            time.Time  `json:"created_at"`
	ProcessedAt          *time.Time `json:"processed_at,omitempty"`
}

// SlotPricing for category queues
type SlotPricing struct {
	ID             uuid.UUID `json:"id"`
	CategoryID     uuid.UUID `json:"category_id"`
	TownID         uuid.UUID `json:"town_id"`
	SkipQueuePrice float64   `json:"skip_queue_price"`
	ExtraSlotPrice float64   `json:"extra_slot_price"`
	IsActive       bool      `json:"is_active"`

	// Joined
	Category *Category `json:"category,omitempty"`
	Town     *Town     `json:"town,omitempty"`
}

// =============================================================================
// REPUTATION & USER EXTENSION MODELS
// =============================================================================

// UserReputation extended user info
type UserReputation struct {
	UserID                 uuid.UUID `json:"user_id"`
	Username               string    `json:"username"`
	FullName               string    `json:"full_name"`
	AvatarURL              *string   `json:"avatar_url,omitempty"`
	Rating                 float64   `json:"rating"`
	RatingCount            int       `json:"rating_count"`
	CompletedAuctions      int       `json:"completed_auctions"`
	BadgeLevel             string    `json:"badge_level"` // none, bronze, silver, gold
	IsTrustedSeller        bool      `json:"is_trusted_seller"`
	IsVerified             bool      `json:"is_verified"`
	IsFastResponder        bool      `json:"is_fast_responder"`
	CompletionRate         float64   `json:"completion_rate"` // % without disputes
	TotalTransactions      int       `json:"total_transactions"`
	SuccessfulTransactions int       `json:"successful_transactions"`
	MemberSince            time.Time `json:"member_since"`
	Badges                 []string  `json:"badges"`
}

// UserRating represents a review
type UserRating struct {
	ID                  uuid.UUID  `json:"id"`
	RaterID             uuid.UUID  `json:"rater_id"`
	RatedUserID         uuid.UUID  `json:"rated_user_id"`
	AuctionID           *uuid.UUID `json:"auction_id,omitempty"`
	Rating              int        `json:"rating"` // 1-5
	CommunicationRating *int       `json:"communication_rating,omitempty"`
	AccuracyRating      *int       `json:"accuracy_rating,omitempty"`
	SpeedRating         *int       `json:"speed_rating,omitempty"`
	Review              *string    `json:"review,omitempty"`
	Role                string     `json:"role"` // buyer, seller
	WouldRecommend      bool       `json:"would_recommend"`
	CreatedAt           time.Time  `json:"created_at"`

	// Joined
	Rater *User `json:"rater,omitempty"`
}

// CreateRatingRequest for rating a user
type CreateRatingRequest struct {
	Rating              int     `json:"rating" binding:"required,min=1,max=5"`
	CommunicationRating *int    `json:"communication_rating" binding:"omitempty,min=1,max=5"`
	AccuracyRating      *int    `json:"accuracy_rating" binding:"omitempty,min=1,max=5"`
	SpeedRating         *int    `json:"speed_rating" binding:"omitempty,min=1,max=5"`
	Review              *string `json:"review"`
	WouldRecommend      bool    `json:"would_recommend"`
}

// =============================================================================
// TOWN LEADERBOARD MODELS
// =============================================================================

// LeaderboardEntry for town rankings
type LeaderboardEntry struct {
	Rank        int       `json:"rank"`
	UserID      uuid.UUID `json:"user_id"`
	Username    string    `json:"username"`
	FullName    string    `json:"full_name"`
	AvatarURL   *string   `json:"avatar_url,omitempty"`
	BadgeLevel  string    `json:"badge_level"`
	Score       float64   `json:"score"`
	MetricValue int       `json:"metric_value"`
}

// TownLeaderboard contains leaderboard data
type TownLeaderboard struct {
	TownID          uuid.UUID          `json:"town_id"`
	TownName        string             `json:"town_name"`
	LeaderboardType string             `json:"leaderboard_type"`
	Period          string             `json:"period"`
	Entries         []LeaderboardEntry `json:"entries"`
	CalculatedAt    time.Time          `json:"calculated_at"`
}

// TownStats aggregated town statistics
type TownStats struct {
	TownID          uuid.UUID  `json:"town_id"`
	TownName        string     `json:"town_name"`
	ActiveAuctions  int        `json:"active_auctions"`
	TotalAuctions   int        `json:"total_auctions"`
	TotalUsers      int        `json:"total_users"`
	ActiveSellers   int        `json:"active_sellers"`
	TotalSalesValue float64    `json:"total_sales_value"`
	AvgAuctionPrice float64    `json:"avg_auction_price"`
	TopCategoryID   *uuid.UUID `json:"top_category_id,omitempty"`
	TopCategory     *Category  `json:"top_category,omitempty"`
	CalculatedAt    time.Time  `json:"calculated_at"`
}

// =============================================================================
// FRAUD DETECTION MODELS
// =============================================================================

// FraudSignal represents a detected suspicious activity
type FraudSignal struct {
	ID          uuid.UUID  `json:"id"`
	UserID      uuid.UUID  `json:"user_id"`
	AuctionID   *uuid.UUID `json:"auction_id,omitempty"`
	SignalType  string     `json:"signal_type"` // self_bidding, rapid_bids, ip_reuse, etc
	Severity    string     `json:"severity"`    // low, medium, high, critical
	Details     string     `json:"details"`     // JSON details
	ScoreImpact float64    `json:"score_impact"`
	IPAddress   *string    `json:"ip_address,omitempty"`
	IsReviewed  bool       `json:"is_reviewed"`
	ReviewedBy  *uuid.UUID `json:"reviewed_by,omitempty"`
	ReviewedAt  *time.Time `json:"reviewed_at,omitempty"`
	AutoFlagged bool       `json:"auto_flagged"`
	CreatedAt   time.Time  `json:"created_at"`
}

// UserBehaviorMetrics for fraud detection
type UserBehaviorMetrics struct {
	UserID              uuid.UUID `json:"user_id"`
	BidCancelRate       float64   `json:"bid_cancel_rate"`
	AvgBidTimeBeforeEnd *int      `json:"avg_bid_time_before_end_mins,omitempty"`
	SameIPBiddersCount  int       `json:"same_ip_bidders_count"`
	RapidBidCount       int       `json:"rapid_bid_count"`
	SelfBidAttempts     int       `json:"self_bid_attempts"`
	ShillBidProbability float64   `json:"shill_bid_probability"`
	AccountAgeDays      int       `json:"account_age_days"`
	RiskLevel           string    `json:"risk_level"` // low, medium, high, critical
	LastCalculatedAt    time.Time `json:"last_calculated_at"`
}
